File: scripts/import-wordpress.py
#!/usr/bin/env python3
import os
import sys
import xml.etree.ElementTree as ET
import re
import requests
from datetime import datetime
import json
from pathlib import Path

# Sanity credentials from environment
PROJECT_ID = os.getenv('SANITY_PROJECT_ID', 'p1uw1wga')
DATASET = os.getenv('SANITY_DATASET', 'production')
API_TOKEN = os.getenv('SANITY_API_TOKEN')

print(f"[v0] Starting WordPress import...")
print(f"[v0] Project ID: {PROJECT_ID}")
print(f"[v0] Dataset: {DATASET}")
print(f"[v0] Token present: {bool(API_TOKEN)}")

# Try multiple possible paths for the XML file
# First try environment variable
xml_file = os.getenv('WORDPRESS_XML_PATH', '')

if not xml_file or not os.path.exists(xml_file):
    possible_paths = [
        '/vercel/share/v0-project/scripts/wordpress-export.xml',
        './wordpress-export.xml',
        '../scripts/wordpress-export.xml',
        './scripts/wordpress-export.xml',
        os.path.expanduser('~/wordpress-export.xml')
    ]
    
    xml_file = None
    for path in possible_paths:
        if os.path.exists(path):
            xml_file = path
            print(f"[v0] Found XML file at: {xml_file}")
            break

if not xml_file:
    print(f"[v0] Error: Could not find wordpress-export.xml")
    print(f"[v0] Tried: {possible_paths}")
    sys.exit(1)

def html_to_blocks(html):
    """Convert HTML to Sanity portable text blocks"""
    if not html:
        return []
    
    # Remove script and style tags
    html = re.sub(r'<script[^>]*>.*?</script>', '', html, flags=re.DOTALL)
    html = re.sub(r'<style[^>]*>.*?</style>', '', html, flags=re.DOTALL)
    
    # Split by paragraphs
    blocks = []
    paragraphs = re.split(r'</?p[^>]*>', html)
    
    for para in paragraphs:
        para = para.strip()
        if not para:
            continue
        
        # Remove HTML tags for now (basic conversion)
        text = re.sub(r'<[^>]+>', '', para)
        text = text.strip()
        
        if text:
            blocks.append({
                "_type": "block",
                "style": "normal",
                "children": [
                    {
                        "_type": "span",
                        "text": text
                    }
                ]
            })
    
    return blocks

def download_image(url):
    """Download image and return asset reference"""
    try:
        if not url or not url.startswith('http'):
            return None
        
        response = requests.get(url, timeout=10)
        if response.status_code == 200:
            # Upload to Sanity
            files = {'file': ('image.jpg', response.content, 'image/jpeg')}
            upload_url = f"https://api.sanity.io/v1/uploads/images/{PROJECT_ID}/{DATASET}"
            headers = {'Authorization': f'Bearer {API_TOKEN}'}
            
            upload_response = requests.post(upload_url, files=files, headers=headers)
            if upload_response.status_code == 201:
                asset_data = upload_response.json()
                return {
                    "_type": "image",
                    "asset": {
                        "_type": "reference",
                        "_ref": asset_data['document']['_id']
                    }
                }
    except Exception as e:
        print(f"[v0] Error downloading image {url}: {e}")
    
    return None

try:
    tree = ET.parse(xml_file)
    root = tree.getroot()
    print(f"[v0] XML parsed successfully")
    
    # Define namespaces
    namespaces = {
        'wp': 'http://wordpress.org/export/1.2/',
        'content': 'http://purl.org/rss/1.0/modules/content/',
        'excerpt': 'http://purl.org/rss/1.0/modules/excerpt/'
    }
    
    items = root.findall('.//item')
    print(f"[v0] Found {len(items)} total items in XML")
    
    posts_to_import = []
    post_count = 0
    
    for item in items:
        # Check if it's a post (not attachment, page, etc)
        post_type_elem = item.find('wp:post_type', namespaces)
        if post_type_elem is None or post_type_elem.text != 'post':
            continue
        
        # Check if post is published
        status_elem = item.find('wp:status', namespaces)
        if status_elem is None or status_elem.text != 'publish':
            continue
        
        post_count += 1
        
        # Extract post data
        title_elem = item.find('title')
        title = title_elem.text if title_elem is not None else f"Post {post_count}"
        
        link_elem = item.find('link')
        link = link_elem.text if link_elem is not None else ""
        slug = link.split('/')[-2] if link else f"post-{post_count}"
        
        pub_date_elem = item.find('pubDate')
        date_str = pub_date_elem.text if pub_date_elem is not None else datetime.now().isoformat()
        
        content_elem = item.find('content:encoded', namespaces)
        content = content_elem.text if content_elem is not None else ""
        
        excerpt_elem = item.find('excerpt:encoded', namespaces)
        excerpt = excerpt_elem.text if excerpt_elem is not None else ""
        
        # Create Sanity document
        doc = {
            "_type": "blogPost",
            "title": title,
            "slug": {
                "_type": "slug",
                "current": slug
            },
            "date": date_str,
            "excerpt": excerpt if excerpt else "",
            "content": html_to_blocks(content)
        }
        
        posts_to_import.append(doc)
        print(f"[v0] Post {post_count}: {title}")
        print(f"[v0]   Slug: {slug}")
        print(f"[v0]   Date: {date_str}")
        print(f"[v0]   Content blocks: {len(doc['content'])}")
    
    print(f"\n[v0] Total posts ready for import: {post_count}")
    
    # Import to Sanity
    if posts_to_import and API_TOKEN:
        print(f"[v0] Importing {len(posts_to_import)} posts to Sanity...")
        
        url = f"https://api.sanity.io/v1/data/mutate/{PROJECT_ID}"
        headers = {
            'Authorization': f'Bearer {API_TOKEN}',
            'Content-Type': 'application/json'
        }
        
        mutations = [
            {
                "createOrReplace": post
            }
            for post in posts_to_import
        ]
        
        payload = {
            "mutations": mutations
        }
        
        response = requests.post(url, json=payload, headers=headers)
        
        if response.status_code == 200:
            result = response.json()
            print(f"[v0] Successfully imported {len(posts_to_import)} posts!")
            print(f"[v0] Response: {json.dumps(result, indent=2)}")
        else:
            print(f"[v0] Error importing to Sanity: {response.status_code}")
            print(f"[v0] Response: {response.text}")
    else:
        print(f"[v0] No posts to import or API token missing")
    
except Exception as e:
    print(f"[v0] Error: {e}")
    import traceback
    traceback.print_exc()

File: scripts/import-wordpress.js
const { createClient } = require("@sanity/client");
const { XMLParser } = require("fast-xml-parser");
const { readFileSync } = require("fs");
const { resolve, dirname } = require("path");

const PROJECT_ID = "z9la8qla";
const DATASET = "production";
const API_VERSION = "2024-01-01";

const token = process.env.SANITY_API_TOKEN;
if (!token) {
  console.error("ERROR: SANITY_API_TOKEN environment variable is not set.");
  console.error("Add it in your project's Vars settings and try again.");
  process.exit(1);
}

const sanity = createClient({
  projectId: PROJECT_ID,
  dataset: DATASET,
  apiVersion: API_VERSION,
  token,
  useCdn: false,
});

// ─── HTML → Portable Text converter ──────────────────────────────────────────

function stripWpBlockComments(html) {
  return html.replace(/<!-- wp:[^>]*-->/g, "").replace(/<!-- \/wp:[^>]*-->/g, "");
}

function extractFirstImageUrl(html) {
  const match = html.match(/<img[^>]+src=["']([^"'?]+)/);
  return match ? match[1] : null;
}

function htmlToPortableText(html) {
  if (!html) return [];

  const cleaned = stripWpBlockComments(html);
  const blocks = [];

  const segments = cleaned
    .split(/(?=<(?:p|h[1-6]|figure|ul|ol|blockquote)[^>]*>)|(?<=<\/(?:p|h[1-6]|figure|ul|ol|blockquote)>)/g);

  for (const seg of segments) {
    const trimmed = seg.trim();
    if (!trimmed) continue;

    if (trimmed.includes("wp-block-jetpack-contact-form")) continue;

    const imgMatch = trimmed.match(/<img[^>]+src=["']([^"'?]+)[^>]*alt=["']([^"']*)/);
    if (imgMatch) continue;

    const headingMatch = trimmed.match(/<h([1-6])[^>]*>(.*?)<\/h[1-6]>/is);
    if (headingMatch) {
      const level = parseInt(headingMatch[1]);
      const style = level <= 2 ? "h2" : level === 3 ? "h3" : "h4";
      blocks.push({
        _type: "block",
        _key: randomKey(),
        style,
        children: [{ _type: "span", _key: randomKey(), text: stripTags(headingMatch[2]), marks: [] }],
        markDefs: [],
      });
      continue;
    }

    const bqMatch = trimmed.match(/<blockquote[^>]*>(.*?)<\/blockquote>/is);
    if (bqMatch) {
      blocks.push({
        _type: "block",
        _key: randomKey(),
        style: "blockquote",
        children: [{ _type: "span", _key: randomKey(), text: stripTags(bqMatch[1]), marks: [] }],
        markDefs: [],
      });
      continue;
    }

    const pMatch = trimmed.match(/<p[^>]*>(.*?)<\/p>/is);
    if (pMatch) {
      const inner = pMatch[1].trim();
      if (!inner || inner === "&nbsp;") continue;
      const { children, markDefs } = parseInlineHtml(inner);
      if (children.length > 0) {
        blocks.push({
          _type: "block",
          _key: randomKey(),
          style: "normal",
          children,
          markDefs,
        });
      }
      continue;
    }

    const bare = stripTags(trimmed).trim();
    if (bare && bare !== "&nbsp;" && !bare.includes("<!--")) {
      blocks.push({
        _type: "block",
        _key: randomKey(),
        style: "normal",
        children: [{ _type: "span", _key: randomKey(), text: bare, marks: [] }],
        markDefs: [],
      });
    }
  }

  return blocks;
}

function parseInlineHtml(html) {
  const markDefs = [];
  const children = [];

  let processed = html
    .replace(/<strong>(.*?)<\/strong>/gis, (_, t) => `__STRONG__${t}__/STRONG__`)
    .replace(/<b>(.*?)<\/b>/gis, (_, t) => `__STRONG__${t}__/STRONG__`)
    .replace(/<em>(.*?)<\/em>/gis, (_, t) => `__EM__${t}__/EM__`)
    .replace(/<i>(.*?)<\/i>/gis, (_, t) => `__EM__${t}__/EM__`);

  processed = processed.replace(/<a\s+href=["']([^"']+)["'][^>]*>(.*?)<\/a>/gis, (_, href, text) => {
    const key = randomKey();
    markDefs.push({ _key: key, _type: "link", href });
    return `__LINK:${key}__${text}__/LINK__`;
  });

  processed = processed.replace(/<[^>]+>/g, "");
  processed = processed.replace(/&amp;/g, "&").replace(/&nbsp;/g, " ").replace(/&lt;/g, "<").replace(/&gt;/g, ">");

  const tokens = processed.split(/(__(?:STRONG|EM|LINK:[^_]+)__|__\/(?:STRONG|EM|LINK)__)/);
  let activeBold = false;
  let activeItalic = false;
  let activeLinkKey = null;

  for (const token of tokens) {
    if (token === "__STRONG__") { activeBold = true; continue; }
    if (token === "__/STRONG__") { activeBold = false; continue; }
    if (token === "__EM__") { activeItalic = true; continue; }
    if (token === "__/EM__") { activeItalic = false; continue; }
    if (token === "__/LINK__") { activeLinkKey = null; continue; }
    const linkOpen = token.match(/^__LINK:([^_]+)__$/);
    if (linkOpen) { activeLinkKey = linkOpen[1]; continue; }

    if (!token) continue;

    const marks = [];
    if (activeBold) marks.push("strong");
    if (activeItalic) marks.push("em");
    if (activeLinkKey) marks.push(activeLinkKey);

    children.push({ _type: "span", _key: randomKey(), text: token, marks });
  }

  return { children, markDefs };
}

function stripTags(html) {
  return html
    .replace(/<[^>]+>/g, "")
    .replace(/&amp;/g, "&")
    .replace(/&nbsp;/g, " ")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .trim();
}

function randomKey() {
  return Math.random().toString(36).slice(2, 10);
}

// ─── Upload image from URL ────────────────────────────────────────────────────

async function uploadImageFromUrl(imageUrl) {
  try {
    const cleanUrl = imageUrl.split("?")[0];
    console.log(`  Uploading image: ${cleanUrl}`);

    const response = await fetch(cleanUrl);
    if (!response.ok) {
      console.warn(`  Could not fetch image (${response.status}): ${cleanUrl}`);
      return null;
    }

    const contentType = response.headers.get("content-type") || "image/jpeg";
    const buffer = Buffer.from(await response.arrayBuffer());
    const ext = contentType.includes("png") ? "png" : contentType.includes("gif") ? "gif" : "jpg";

    const asset = await sanity.assets.upload("image", buffer, {
      filename: `wp-import-${randomKey()}.${ext}`,
      contentType,
    });

    return { _type: "image", asset: { _type: "reference", _ref: asset._id } };
  } catch (err) {
    console.warn(`  Image upload failed: ${err.message}`);
    return null;
  }
}

// ─── Parse XML ────────────────────────────────────────────────────────────────

function parseXml(xmlPath) {
  const xml = readFileSync(xmlPath, "utf-8");
  const parser = new XMLParser({
    ignoreAttributes: false,
    attributeNamePrefix: "@_",
    cdataPropName: "__cdata",
    isArray: (name) => ["item", "category", "wp:postmeta"].includes(name),
  });
  const result = parser.parse(xml);
  return result?.rss?.channel?.item || [];
}

// ─── Main ─────────────────────────────────────────────────────────────────────

async function main() {
  const xmlPath = resolve(__dirname, "wordpress-export.xml");

  console.log("Parsing WordPress XML export...");
  const items = parseXml(xmlPath);

  const posts = items.filter((item) => {
    const type = item["wp:post_type"]?.__cdata || item["wp:post_type"];
    const status = item["wp:status"]?.__cdata || item["wp:status"];
    return type === "post" && status === "publish";
  });

  console.log(`Found ${posts.length} published posts to import.\n`);

  for (const item of posts) {
    const title = item.title?.__cdata || item.title || "Untitled";
    const slug = item["wp:post_name"]?.__cdata || item["wp:post_name"] || "";
    const rawDate = item["wp:post_date_gmt"]?.__cdata || item["wp:post_date_gmt"] || "";
    const date = rawDate ? rawDate.split(" ")[0] : new Date().toISOString().split("T")[0];
    const htmlContent = item["content:encoded"]?.__cdata || "";

    console.log(`Importing: "${title}"`);

    const existing = await sanity.fetch(
      `*[_type == "blogPost" && slug.current == $slug][0]._id`,
      { slug }
    );
    if (existing) {
      console.log(`  Skipping — already exists (${slug})\n`);
      continue;
    }

    let thumbnail = null;
    const firstImgUrl = extractFirstImageUrl(htmlContent);
    if (firstImgUrl) {
      thumbnail = await uploadImageFromUrl(firstImgUrl);
    }

    const content = htmlToPortableText(htmlContent);

    const firstParagraph = htmlContent.match(/<p[^>]*>(.*?)<\/p>/is)?.[1] || "";
    const excerpt = stripTags(firstParagraph).slice(0, 280) || "";

    const doc = {
      _type: "blogPost",
      title,
      slug: { _type: "slug", current: slug },
      date,
      excerpt: excerpt || undefined,
      content: content.length > 0 ? content : undefined,
      ...(thumbnail ? { thumbnail } : {}),
    };

    try {
      const created = await sanity.create(doc);
      console.log(`  Created: ${created._id}\n`);
    } catch (err) {
      console.error(`  ERROR creating "${title}": ${err.message}\n`);
    }
  }

  console.log("Import complete!");
}

main().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});
